### ISSUE TITLE
Linker warnings IL2008 and IL2009 when trimming non-Windows CoreCLR application

### ISSUE THREAD

**Issue by eerhardt** [dotnet/runtime#40336]
When publishing a linux-x64 app with

```xml
    <PublishTrimmed>true</PublishTrimmed>
    <TrimMode>link</TrimMode>
```

I'm getting the following warnings that I don't get on a Windows CoreCLR application:

```
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2008: Could not resolve type 'System.Runtime.InteropServices.IDispatch' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2008: Could not resolve type 'System.Runtime.InteropServices.IMarshal' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2009: Could not find method 'ForwardCallToInvokeMember' in type 'System.RuntimeType' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2008: Could not resolve type 'System.__ComObject' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2009: Could not find method 'SetValue' in type 'System.Reflection.RuntimeFieldInfo' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2009: Could not find method 'GetValue' in type 'System.Reflection.RuntimeFieldInfo' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2008: Could not resolve type 'System.Variant' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2009: Could not find method 'CallComputeVtables' in type 'System.Runtime.InteropServices.ComWrappers' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2009: Could not find method 'CallCreateObject' in type 'System.Runtime.InteropServices.ComWrappers' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2009: Could not find method 'CallReleaseObjects' in type 'System.Runtime.InteropServices.ComWrappers' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2009: Could not find method 'CallICustomQueryInterface' in type 'System.Runtime.InteropServices.ComWrappers' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2008: Could not resolve type 'System.Runtime.InteropServices.ComWrappersScenario' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2008: Could not resolve type 'System.OleAutBinder' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2009: Could not find method 'GetCOMHRExceptionObject' in type 'System.StubHelpers.StubHelpers' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2009: Could not find method 'GetCOMIPFromRCW' in type 'System.StubHelpers.StubHelpers' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2008: Could not resolve type 'System.StubHelpers.ObjectMarshaler' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2008: Could not resolve type 'System.StubHelpers.InterfaceMarshaler' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2008: Could not resolve type 'System.StubHelpers.MngdSafeArrayMarshaler' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e [F:\DotNetTest\HelloWorld\HelloWorld.csproj]
```

This is because we are referencing Windows-only types in our ILLink.Descriptors.xml file for all platforms of CoreCLR.

https://github.com/dotnet/runtime/blob/c91269cb959a78309d0fa3040ae6028cce97f86e/src/coreclr/src/System.Private.CoreLib/ILLinkTrim.xml#L17-L18

And also when we generate the XML file (using [CreateRuntimeRootDescriptorFile](https://github.com/mono/linker/blob/master/src/ILLink.Tasks/CreateRuntimeRootDescriptorFile.cs)), we aren't excluding the types that are guarded with `#ifdef FEATURE_COMINTEROP`. For example:

https://github.com/dotnet/runtime/blob/29e9b5b7fd95231d9cd9d3ae351404e63cbb6d5a/src/coreclr/src/vm/corelib.h#L1059-L1078

In order to fix this, we should ensure we don't add linker XML entries for types that don't exist on other platforms.

cc @sbomer @vitek-karas

**Comment by eerhardt** [comment]
We are also getting `IL2037` when linking the runtimepack on non-Windows.

```
/_/src/libraries/System.Private.CoreLib/src/System/AppDomain.cs(412,13): error IL2037: System.AppDomain.GetThreadPrincipal(): No members were resolved for 'GetDefaultInstance'. [/Users/runner/work/1/s/src/libraries/src.proj] 
##[error]/_/src/libraries/System.Private.CoreLib/src/System/AppDomain.cs(412,13): error IL2037: (NETCORE_ENGINEERING_TELEMETRY=Build) System.AppDomain.GetThreadPrincipal(): No members were resolved for 'GetDefaultInstance'. 
/_/src/libraries/System.Private.CoreLib/src/System/AppDomain.cs(412,13): error IL2037: System.AppDomain.GetThreadPrincipal(): No members were resolved for 'GetDefaultInstance'. [/Users/runner/work/1/s/src/libraries/src.proj] 
##[error]/_/src/libraries/System.Private.CoreLib/src/System/AppDomain.cs(412,13): error IL2037: (NETCORE_ENGINEERING_TELEMETRY=Build) System.AppDomain.GetThreadPrincipal(): No members were resolved for 'GetDefaultInstance'.
```

See the global suppression added in https://github.com/dotnet/runtime/pull/40268/commits/a1a8b489c737b6afbebe9157571e25357dc36974#diff-749a6d32ee50a25c6076a389e2c805e3R27. We need to fix that as well.

**Comment by eerhardt** [comment]
Looks like we will also need to respect `32-bit` vs `64-bit` as well, since we are getting the following warnings on the x86 build:

```
resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e(0,0): error IL2012: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not find field 'Padding' in type 'System.Runtime.CompilerServices.RawArrayData' specified in resource System.Private.CoreLib.xml in System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
```

**Comment by eerhardt** [comment]
This was resolved with #52457 (and other work before it).
